yang/frr-bgp.yang

```diff
@@ -0,0 +1,1248 @@
+
+  include "frr-bgp-neighbor";
+
+  include "frr-bgp-peer-group";
+
+  include "frr-bgp-bmp";
+
+  organization
+    "FRRouting";
+  contact
+    "FRR Users List: <mailto:frog@lists.frrouting.org> FRR Development
+     List: <mailto:dev@lists.frrouting.org>";
+  description
+    "This module defines a model for managing FRR bgpd daemon.
+
+     Copyright 2020 FRRouting
+
+     Redistribution and use in source and binary forms, with or without
+     modification, are permitted provided that the following conditions
+     are met:
+
+     1. Redistributions of source code must retain the above copyright notice,
+     this list of conditions and the following disclaimer.
+
+     2. Redistributions in binary form must reproduce the above copyright
+     notice, this list of conditions and the following disclaimer in the
+     documentation and/or other materials provided with the distribution.
+
+     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+     \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+     LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+     A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+     HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+     SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+     LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+     DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+     THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+     (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+     OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.";
+
+  revision 2019-12-03 {
+    description
+      "Initial revision.";
+  }
+
+  identity bgp {
+    base frr-rt:routing-protocol;
+    description
+      "BGP protocol.";
+  }
+
+  grouping mp-afi-unicast-common {
+    uses global-group-use-multiple-paths;
+
+    uses global-redistribute;
+
+    uses admin-distance;
+  }
+
+  augment "/frr-rt:routing/frr-rt:control-plane-protocols/frr-rt:control-plane-protocol" {
+    container bgp {
+      when "../frr-rt:type = 'frr-bgp:bgp'" {
+        description
+          "BGP protocol augmentation of ietf-routing module
+           control-plane-protocol.";
+      }
+      description
+        "Top-level configuration for the BGP router.";
+      container global {
+        presence "Enables global configuration of BGP";
+        description
+          "Global configuration for the BGP router.";
+        leaf local-as {
+          type inet:as-number;
+          mandatory true;
+          description
+            "Local autonomous system number of the router.  Uses
+             the 32-bit as-number type from the model in RFC 6991.";
+        }
+
+        leaf instance-type {
+          type frr-bt:bgp-instance-type;
+          mandatory true;
+          description
+            "BGP instance type";
```

Comment:
I think this should be made RO, and then we should add a separate RW boolean leaf that specifies whether this instance is a `view` or not. Reason being is that if we are in a non-default VRF, then we already know it's a VRF instance, so that doesn't need to be explicitly configured. The other two cases - view or "regular" - are determined by the view-ness, which can itself be set or unset.

Reply:
I have introduced the new leaf instance-type-view for the same. I am keeping the typedef : bgp-instance-type, we may use this type for operational model.